python/ingestion/gcs_to_bq_util.py

```diff
@@ -30,22 +33,62 @@ def append_dataframe_to_bq(
     else:
         job_config.schema = get_schema(frame, column_types, col_modes)
 
-    # Add an upload timestamp. Formatting to a string helps BQ autodetection.
-    frame['ingestion_ts'] = datetime.now(
-        timezone.utc).strftime("%Y-%m-%d %H:%M:%S.%f %Z")
+    return job_config
+
+
+def __append_dataframe_to_bq(frame, dataset, table_name, column_types, col_modes, project, json_data):
+    job_config = __create_bq_load_job_config(frame, column_types, col_modes)
 
     client = bigquery.Client(project)
     table_id = client.dataset(dataset).table(table_name)
-    # Repeated fields are not supported with bigquery.Client.load_table_from_dataframe()
-    # (See https://github.com/googleapis/python-bigquery/issues/19). We have to
-    # use load_table_from_json as a workaround.
-    result = frame.to_json(orient='records')
-    json_data = json.loads(result)
+
     load_job = client.load_table_from_json(
         json_data,	table_id, job_config=job_config)
     load_job.result()  # Wait for table load to complete.
 
 
+def append_dataframe_to_bq_as_str_values(frame, dataset, table_name, column_types=None, col_modes=None, project=None):
+    """Appends the provided DataFrame to the table specified by
+       `dataset.table_name`. Automatically adds an ingestion time column and coverts
+       all other values to a string.
+
+       frame: pandas.DataFrame representing the data to append.
+       dataset: The BigQuery dataset to write to.
+       table_name: The BigQuery table to write to.
+       column_types: Optional dict of column name to BigQuery data type. If
+                     present, the column names must match the columns in the
+                     DataFrame. Otherwise, table schema is inferred.
+       col_modes: Optional dict of modes for each field. Possible values include
+                  NULLABLE, REQUIRED, and REPEATED. Must also specify
+                  column_types to specify col_modes."""
+    json_data = __convert_frame_to_json(frame, column_types)
+    for sub in json_data:
+        for key in sub:
+            sub[key] = str(sub[key])
```

Comment:
Can you use DataFrame.convert_dtypes(convert_string=True) before converting to json instead? Then it could be taken care of in the ManualUploads class instead. Another option could be DataFrame.applymap(str)

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.convert_dtypes.html

Reply:
Nice find, I gave this a try but no luck. I've tried similar approaches but the only way that this works is to do the string casting after converting to json.

Reply:
Bummer :/ thanks for trying it!